src/flp/types.rs

```diff
@@ -471,6 +472,234 @@ where
+        g: &mut Vec<Box<dyn Gadget<F>>>,
+        input: &[F],
+        joint_rand: &[F],
+        num_shares: usize,
+    ) -> Result<F, FlpError> {
+        self.valid_call_check(input, joint_rand)?;
+
+        // Check that each element of `input` is a 0 or 1.
+        let range_check = parallel_sum_range_checks(
+            &mut g[0],
+            input,
+            joint_rand[0],
+            self.chunk_length,
+            num_shares,
+        )?;
+
+        // Check that the elements of `input` sum to at most `max_weight`.
+        let count_vec = &input[..self.length];
+        let weight = count_vec.iter().fold(F::zero(), |a, b| a + *b);
+        let offset_weight_reported = F::decode_bitvector(&input[self.length..])?;
+
+        // From spec: weight_check = self.offset*shares_inv + weight - weight_reported
+        let weight_check = {
+            let bits_for_weight = ilog2(self.max_weight) as usize + 1;
+            let offset = F::from(F::valid_integer_try_from(
+                (1 << bits_for_weight) - 1 - self.max_weight,
+            )?);
+
+            let shares_inv = F::from(F::valid_integer_try_from(num_shares)?).inv();
+            offset * shares_inv + weight - offset_weight_reported
+        };
+
+        // Take a random linear combination of both checks.
+        let out = joint_rand[1] * range_check + (joint_rand[1] * joint_rand[1]) * weight_check;
+        Ok(out)
+    }
+
+    // Truncates the measurement, removing extra data that was necessary for validity (here, the
+    // encoded weight), but not important for aggregation
+    fn truncate(&self, input: Vec<Self::Field>) -> Result<Vec<Self::Field>, FlpError> {
+        // Cut off the encoded weight
+        Ok(input[..self.length].to_vec())
+    }
+
+    // The length in field elements of the encoded input returned by [`Self::encode_measurement`].
+    fn input_len(&self) -> usize {
+        let bits_for_weight = ilog2(self.max_weight) as usize + 1;
+        self.length + bits_for_weight
+    }
+
+    fn proof_len(&self) -> usize {
+        // TODO: where does this calculation come from? I copied it from Histogram
+        (self.chunk_length * 2) + 2 * ((1 + self.gadget_calls).next_power_of_two() - 1) + 1
+    }
+
+    fn verifier_len(&self) -> usize {
+        // TODO: where does this calculation come from? I copied it from Histogram
+        2 + self.chunk_length * 2
+    }
+
+    // The length of the truncated output (i.e., the output of [`Type::truncate`]).
+    fn output_len(&self) -> usize {
+        self.length
+    }
+
+    // The number of random values needed in the validity checks
+    fn joint_rand_len(&self) -> usize {
+        2
+    }
+
+    fn prove_rand_len(&self) -> usize {
+        // TODO: where does this calculation come from? I copied it from Histogram
+        self.chunk_length * 2
+    }
+
+    fn query_rand_len(&self) -> usize {
+        // TODO: where does this calculation come from? I copied it from Histogram
+        1
+    }
```

Comment:
These are similarly derived from the same block of methods, with the gadget details substituted in an simplified.

However, note that we can't yet implement Prio3MultihotCountVec exactly as in draft-12 yet, since this codebase is still on draft-08, and hasn't adopted the `EVAL_OUTPUT_LEN` change yet, so we have to do a bit of backporting. Thus, we'll keep `query_rand_len` equal to 1, for the one evaluation point we need for one gadget polynomial test. Once we get up to draft-10, then this and other VDAFs will need more query randomness values, in order to compute a random linear combination of their circuit outputs.

Reply:
Gotcha I'll make a note